File: buerguer-pig/AMarotoChivite-BurguerPig/AMarotoChivite-BurguerPig-Java/src/main/java/Ficheros/BurguerPig/controllers/ControllerQueries.java
package Ficheros.BurguerPig.controllers;

import Ficheros.BurguerPig.models.Burguer;
import Ficheros.BurguerPig.models.Ingredient;
import Ficheros.BurguerPig.repositories.RepositoryIngredient;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ControllerQueries {

    private final List<Burguer> burguerListOfStorageReadFile;
    private final RepositoryIngredient repoIngredients;

    private final Map<String, List<Ingredient>> groupByIngredient;

    public ControllerQueries(List<Burguer> burguerListOfStorageReadFile) {
        this.burguerListOfStorageReadFile = burguerListOfStorageReadFile;
        this.repoIngredients = new RepositoryIngredient();

        groupByIngredient = repoIngredients.getAllIngredient().stream()
                .collect(Collectors.groupingBy(Ingredient::getName));
    }

    public Burguer burguerMoreExpensive() {
        return burguerListOfStorageReadFile.stream().max(Comparator.comparing(Burguer::getPrice)).orElse(null);
    }

    public Burguer burguerMoreIngredients() {
        return burguerListOfStorageReadFile.stream().max(Comparator.comparingInt(b -> b.getIngredients().size())).orElse(null);
    }

    public Map<String, Integer> numBurguerByIngredient() {
        Map<String, Integer> result = new HashMap<>();
        groupByIngredient.forEach((ingredientName, ingredientList) -> {
            int count = 0;
            for (Burguer burguer : burguerListOfStorageReadFile) {
                if (burguer.getIngredients().stream().anyMatch(ingredient -> ingredient != null && ingredient.getName().equals(ingredientName))) {
                    count++;
                }
            }
            result.put(ingredientName, count);
        });
        return result;
    }

    public Map<String, List<Burguer>> burguersByIngredient() {
        Map<String, List<Burguer>> result = new HashMap<>();
        groupByIngredient.forEach((ingredientName, ingredientList) -> {
            List<Burguer> burguers = burguerListOfStorageReadFile.stream()
                    .filter(burguer -> burguer.getIngredients().stream()
                            .anyMatch(ingredient -> ingredient != null && ingredient.getName().equals(ingredientName)))
                    .collect(Collectors.toList());
            result.put(ingredientName, burguers);
        });
        return result;
    }

    public double averagePriceBurguers() {
        return burguerListOfStorageReadFile.stream().mapToDouble(Burguer::getPrice).average().orElse(0.0);
    }

    public double averagePriceIngredients() {
        return repoIngredients.getAllIngredient().stream().mapToDouble(Ingredient::getPrice).average().orElse(0.0);
    }
}

File: buerguer-pig/AMarotoChivite-BurguerPig/AMarotoChivite-BurguerPig-Kotlin/src/main/kotlin/Ficheros/BurguerPig/controllers/ControllerQueries.kt
package Ficheros.BurguerPig.controllers

import Ficheros.BurguerPig.models.Burguer
import Ficheros.BurguerPig.repositories.RepositoryIngredient

class ControllerQueries(private val burguerListOfStorageReadFile: List<Burguer>) {

    private val repoIngredients = RepositoryIngredient()

    private val groupByIngredient = repoIngredients.getAllIngredient().groupBy { it.name }

    fun burguerMoreExpensive(): Burguer {
        return burguerListOfStorageReadFile.maxBy { it.getPrice() }
    }

    fun burguerMoreIngredients(): Burguer {
        return burguerListOfStorageReadFile.maxBy { it.ingredients.size }
    }

    fun numBurguerByIngredient(): Map<String, Int> {
        return groupByIngredient.mapValues {
            burguerListOfStorageReadFile.count { burguer ->
                burguer.ingredients.any { ingredient -> ingredient?.name == it.key }
            }
        }
    }

    fun burguersByIngredient(): Map<String, List<Burguer>> {
        return groupByIngredient.mapValues { ingredient ->
            burguerListOfStorageReadFile.filter { burguer ->
                burguer.ingredients.any { it?.name == ingredient.key }
            }
        }
    }

    fun averagePriceBurguers(): Double {
        return burguerListOfStorageReadFile.map { it.getPrice() }.average()
    }

    fun averagePriceIngredients(): Double {
        return repoIngredients.getAllIngredient().map { it.price }.average()
    }

}